File: legacy/acmp.js
/* =========================================================================
 * acmp.js — Apoio à Classificação de Motivos de Parada
 *
 * Responsabilidade: sugerir ao operador os motivos mais prováveis no momento
 * em que ele registra uma parada, e permitir avaliar honestamente se essa
 * sugestão vale alguma coisa.
 *
 * O problema: o tempo parado é medido sozinho, mas a causa é digitada por
 * alguém com a máquina parada e a produção atrasando. Diante de vinte
 * opções, escolhe-se a primeira plausível. O erro não aparece no OEE —
 * aparece no Pareto, que decide onde a manutenção investe.
 *
 * Método: Naive Bayes multinomial sobre atributos categóricos, com
 * suavização de Laplace, treinado com o próprio histórico de apontamentos.
 *
 * Por que um modelo simples, e não o estado da arte:
 *   1. Treina em milissegundos dentro do navegador, sem servidor, sem GPU,
 *      sem etapa de treino offline — condição real de um terminal de fábrica.
 *   2. É explicável por construção. Dá para mostrar ao operador POR QUE a
 *      sugestão apareceu, e sugestão que o operador não entende é sugestão
 *      ignorada. Dois pontos de acurácia valem menos que isso aqui.
 *   3. Funciona com pouco dado, que é a situação de qualquer implantação
 *      nova.
 *
 * TRÊS ARMADILHAS TRATADAS EXPLICITAMENTE (ver comentários no código):
 *   - vazamento pelo estado da máquina;
 *   - uso de atributo indisponível no momento da inferência (duração);
 *   - divisão aleatória em série temporal.
 * ========================================================================= */
(function (global) {
  'use strict';

  var OEE = global.OEE = global.OEE || {};

  /* Amostras mínimas para o módulo se pronunciar. Abaixo disso, a sugestão
   * seria ruído com aparência de conhecimento — pior que não sugerir. */
  var MINIMO_AMOSTRAS = 30;
  var ALFA = 1;              // suavização de Laplace

  /* ---------------------------------------------------------------------
   * 1. Atributos
   *
   * ATRIBUTOS_INICIO é o conjunto disponível no instante em que o operador
   * aperta "registrar parada". Tudo o que ele não sabe nesse momento está
   * fora — inclusive coisas que melhorariam a acurácia medida.
   *
   * Duas exclusões deliberadas:
   *
   * (a) ESTADO DA MÁQUINA. Na aplicação, o estado é derivado do motivo
   *     escolhido (motivo.estadoSugerido). Treinar com ele daria acurácia
   *     altíssima e completamente falsa: o modelo estaria lendo a resposta.
   *     É o caso clássico de vazamento de alvo.
   *
   * (b) DURAÇÃO DA PARADA. Só existe depois que a parada termina. Usá-la no
   *     treino e não tê-la na inferência é o erro silencioso mais comum
   *     neste tipo de trabalho: o número do artigo não se reproduz em campo.
   *     Ela aparece apenas em ATRIBUTOS_EM_CURSO, usado na reclassificação
   *     de uma parada já em andamento — aí o tempo decorrido é conhecido de
   *     fato. A tela de avaliação compara os dois cenários lado a lado.
   * ------------------------------------------------------------------- */
  var ATRIBUTOS_INICIO = ['maquina', 'turno', 'faixaHora', 'produto', 'motivoAnterior'];
  var ATRIBUTOS_EM_CURSO = ATRIBUTOS_INICIO.concat(['faixaDuracao']);

  var ROTULOS_ATRIBUTO = {
    maquina: 'máquina',
    turno: 'turno',
    faixaHora: 'faixa horária',
    produto: 'produto',
    motivoAnterior: 'parada anterior',
    faixaDuracao: 'duração'
  };

  function faixaHoraria(ts) {
    var h = new Date(ts).getHours();
    if (h < 4) return '00h-04h';
    if (h < 8) return '04h-08h';
    if (h < 12) return '08h-12h';
    if (h < 16) return '12h-16h';
    if (h < 20) return '16h-20h';
    return '20h-24h';
  }

  function faixaDuracao(seg) {
    if (seg === null || seg === undefined) return 'desconhecida';
    if (seg <= 60) return 'até 1 min';
    if (seg <= 300) return 'até 5 min';
    if (seg <= 900) return 'até 15 min';
    if (seg <= 3600) return 'até 1 h';
    return 'acima de 1 h';
  }

  /* ---------------------------------------------------------------------
   * 2. Extração de amostras
   *
   * Cada parada já classificada por um humano é uma amostra rotulada. Não
   * há custo de anotação: o rótulo é o trabalho que o operador já faz.
   * ------------------------------------------------------------------- */
  function extrairAmostras(ds, opcoes) {
    opcoes = opcoes || {};
    var ordens = {};
    (ds.ordens || []).forEach(function (o) { ordens[o.id] = o; });

    var paradas = (ds.eventosParada || [])
      .filter(function (p) {
        if (!p.motivoId) return false;                       // sem rótulo, não serve
        if (opcoes.maquinaIds && opcoes.maquinaIds.indexOf(p.maquinaId) < 0) return false;
        if (opcoes.inicio && p.inicio < opcoes.inicio) return false;
        if (opcoes.fim && p.inicio > opcoes.fim) return false;
        return true;
      })
      .sort(function (a, b) { return a.inicio - b.inicio; });

    /* "Motivo da parada anterior na mesma máquina" precisa da ordem
       cronológica e do estado de cada máquina até aquele ponto. */
    var anteriorPorMaquina = {};

    return paradas.map(function (p) {
      var ordem = ordens[p.ordemId];
      var amostra = {
        id: p.id,
        ts: p.inicio,
        maquinaId: p.maquinaId,
        classe: p.motivoId,
        atributos: {
          maquina: p.maquinaId,
          turno: p.turnoId || 'sem-turno',
          faixaHora: faixaHoraria(p.inicio),
          produto: ordem ? ordem.produtoId : 'sem-ordem',
          motivoAnterior: anteriorPorMaquina[p.maquinaId] || 'nenhuma',
          faixaDuracao: faixaDuracao(p.duracaoSeg)
        }
      };
      anteriorPorMaquina[p.maquinaId] = p.motivoId;
      return amostra;
    });
  }

  /* Contexto de inferência: o que se sabe no instante do apontamento. */
  function contextoAtual(ds, maquina, opcoes) {
    opcoes = opcoes || {};
    var agora = opcoes.agora || Date.now();
    var ordem = maquina.ordemAtualId
      ? (ds.ordens || []).filter(function (o) { return o.id === maquina.ordemAtualId; })[0]
      : null;

    /* Última parada encerrada desta máquina — a "parada anterior". */
    var anteriores = (ds.eventosParada || []).filter(function (p) {
      return p.maquinaId === maquina.id && p.motivoId && p.inicio < agora;
    }).sort(function (a, b) { return a.inicio - b.inicio; });
    var anterior = anteriores[anteriores.length - 1];

    var turno = OEE.Model.turnoDoInstante(ds.turnos, agora);

    return {
      maquina: maquina.id,
      turno: turno.turnoId,
      faixaHora: faixaHoraria(agora),
      produto: ordem ? ordem.produtoId : 'sem-ordem',
      motivoAnterior: anterior ? anterior.motivoId : 'nenhuma',
      faixaDuracao: faixaDuracao(opcoes.duracaoSeg === undefined ? null : opcoes.duracaoSeg)
    };
  }

  /* ---------------------------------------------------------------------
   * 3. Treino
   *
   * Naive Bayes: P(classe | atributos) ∝ P(classe) · Π P(atributo | classe),
   * assumindo independência condicional entre atributos. A suposição é
   * falsa aqui (máquina e produto são correlacionados, por exemplo), e
   * ainda assim o classificador costuma acertar o ranking — é o resultado
   * conhecido de que Naive Bayes tolera dependência sem perder a ordenação.
   * ------------------------------------------------------------------- */
  function treinar(amostras, atributos) {
    atributos = atributos || ATRIBUTOS_INICIO;

    var contagemClasse = {};
    var contagem = {};        // contagem[atributo][classe][valor]
    var valores = {};         // conjunto de valores distintos por atributo
    var total = 0;

    atributos.forEach(function (a) { contagem[a] = {}; valores[a] = {}; });

    amostras.forEach(function (am) {
      var c = am.classe;
      contagemClasse[c] = (contagemClasse[c] || 0) + 1;
      total += 1;
      atributos.forEach(function (a) {
        var v = am.atributos[a];
        if (!contagem[a][c]) contagem[a][c] = {};
        contagem[a][c][v] = (contagem[a][c][v] || 0) + 1;
        valores[a][v] = true;
      });
    });

    return {
      atributos: atributos,
      classes: Object.keys(contagemClasse),
      contagemClasse: contagemClasse,
      contagem: contagem,
      cardinalidade: Object.keys(valores).reduce(function (acc, a) {
        acc[a] = Object.keys(valores[a]).length; return acc;
      }, {}),
      total: total,
      treinadoEm: Date.now()
    };
  }

  /* ---------------------------------------------------------------------
   * 4. Inferência
   *
   * Somatório de logaritmos em vez de produto de probabilidades: com cinco
   * atributos e probabilidades pequenas, o produto direto chega a zero por
   * limite de precisão e todas as classes empatam.
   * ------------------------------------------------------------------- */
  function pontuar(modelo, ctx) {
    if (!modelo || !modelo.total) return [];

    var pontos = modelo.classes.map(function (c) {
      var log = Math.log(modelo.contagemClasse[c] / modelo.total);

      modelo.atributos.forEach(function (a) {
        var porClasse = modelo.contagem[a][c] || {};
        var n = porClasse[ctx[a]] || 0;
        var totalClasse = modelo.contagemClasse[c];
        var k = modelo.cardinalidade[a] || 1;
        /* Laplace: valor nunca visto para esta classe recebe probabilidade
           pequena, não zero. Sem isso, um único atributo inédito zeraria a
           classe inteira, por mais evidência que houvesse no resto. */
        log += Math.log((n + ALFA) / (totalClasse + ALFA * k));
      });

      return { classe: c, log: log };
    });

    /* Normalização estável: subtrai o maior log antes de exponenciar. */
    var maior = Math.max.apply(null, pontos.map(function (p) { return p.log; }));
    var soma = 0;
    pontos.forEach(function (p) { p.exp = Math.exp(p.log - maior); soma += p.exp; });
    pontos.forEach(function (p) { p.prob = p.exp / soma; });

    return pontos.sort(function (a, b) { return b.prob - a.prob; });
  }

  /* Evidência legível. Em vez de mostrar P(atributo|classe), que é o que o
   * modelo usa mas ninguém interpreta, mostramos a leitura direta: entre as
   * paradas com este valor de atributo, que fatia foi deste motivo. É a
   * frase que o operador consegue conferir com a própria experiência. */
  function evidencias(modelo, ctx, classe, limite) {
    var lista = modelo.atributos.map(function (a) {
      var valor = ctx[a];
      var comEsseValor = 0;
      var dessaClasse = (modelo.contagem[a][classe] || {})[valor] || 0;
      modelo.classes.forEach(function (c) {
        comEsseValor += (modelo.contagem[a][c] || {})[valor] || 0;
      });
      return {
        atributo: a,
        rotulo: ROTULOS_ATRIBUTO[a] || a,
        valor: valor,
        proporcao: comEsseValor ? dessaClasse / comEsseValor : 0,
        suporte: comEsseValor,
        ocorrencias: dessaClasse
      };
    });

    return lista
      .filter(function (e) { return e.suporte >= 3 && e.proporcao > 0; })
      .sort(function (x, y) { return y.proporcao - x.proporcao; })
      .slice(0, limite || 2);
  }

  function sugerir(modelo, ctx, quantidade) {
    if (!modelo || modelo.total < MINIMO_AMOSTRAS) return [];
    return pontuar(modelo, ctx).slice(0, quantidade || 3).map(function (p, i) {
      return {
        motivoId: p.classe,
        probabilidade: p.prob,
        posicao: i + 1,
        evidencias: evidencias(modelo, ctx, p.classe, 2)
      };
    });
  }

  /* ---------------------------------------------------------------------
   * 5. Referência de comparação
   *
   * Sem baseline, qualquer acurácia parece boa. A régua honesta aqui é o
   * que um sistema sem modelo nenhum já faria: oferecer o motivo mais
   * frequente daquela máquina. Se o classificador não superar isso, ele não
   * paga a própria complexidade.
   * ------------------------------------------------------------------- */
  function baseline(amostrasTreino) {
    var porMaquina = {}, geralMapa = {};
    amostrasTreino.forEach(function (a) {
      porMaquina[a.maquinaId] = porMaquina[a.maquinaId] || {};
      porMaquina[a.maquinaId][a.classe] = (porMaquina[a.maquinaId][a.classe] || 0) + 1;
      geralMapa[a.classe] = (geralMapa[a.classe] || 0) + 1;
    });
    var ordenar = function (mapa) {
      return Object.keys(mapa).sort(function (x, y) { return mapa[y] - mapa[x]; });
    };
    var geralOrdenado = ordenar(geralMapa);
    var porMaq = {}, porMaqTop3 = {};
    Object.keys(porMaquina).forEach(function (m) {
      var lista = ordenar(porMaquina[m]);
      porMaq[m] = lista[0] || null;
      porMaqTop3[m] = lista.slice(0, 3);
    });
    return {
      porMaquina: porMaq,
      porMaquinaTop3: porMaqTop3,
      geral: geralOrdenado[0] || null,
      geralTop3: geralOrdenado.slice(0, 3)
    };
  }

  /* ---------------------------------------------------------------------
   * 6. Avaliação
   *
   * DIVISÃO TEMPORAL, NUNCA ALEATÓRIA. Embaralhar as amostras deixaria o
   * modelo treinar com paradas de quinta e testar com paradas de terça da
   * mesma máquina, no mesmo turno, com o mesmo problema em curso. A acurácia
   * subiria vários pontos e não significaria nada: em produção, o modelo
   * sempre prevê o futuro a partir do passado.
   * ------------------------------------------------------------------- */
  function avaliar(amostras, opcoes) {
    opcoes = opcoes || {};
    var atributos = opcoes.atributos || ATRIBUTOS_INICIO;
    var fracaoTreino = opcoes.fracaoTreino || 0.75;

    if (amostras.length < MINIMO_AMOSTRAS * 2) {
      return { suficiente: false, totalAmostras: amostras.length, minimo: MINIMO_AMOSTRAS * 2 };
    }

    var ordenadas = amostras.slice().sort(function (a, b) { return a.ts - b.ts; });
    var corte = Math.floor(ordenadas.length * fracaoTreino);
    var treino = ordenadas.slice(0, corte);
    var teste = ordenadas.slice(corte);

    var modelo = treinar(treino, atributos);
    var ref = baseline(treino);

    var acertosTop1 = 0, acertosTop3 = 0, acertosBaseline = 0, acertosBaselineTop3 = 0;
    var matriz = {};      // matriz[real][previsto]
    var classesVistas = {};

    teste.forEach(function (am) {
      var ranking = pontuar(modelo, am.atributos);
      var previsto = ranking.length ? ranking[0].classe : null;
      var top3 = ranking.slice(0, 3).map(function (p) { return p.classe; });

      if (previsto === am.classe) acertosTop1 += 1;
      if (top3.indexOf(am.classe) >= 0) acertosTop3 += 1;
      if ((ref.porMaquina[am.maquinaId] || ref.geral) === am.classe) acertosBaseline += 1;
      /* Comparação justa para uma interface que mostra três sugestões:
         os três motivos mais frequentes daquela máquina. Confrontar o
         top-3 do modelo com o top-1 da referência inflaria o ganho. */
      var refTop3 = ref.porMaquinaTop3[am.maquinaId] || ref.geralTop3;
      if (refTop3.indexOf(am.classe) >= 0) acertosBaselineTop3 += 1;

      matriz[am.classe] = matriz[am.classe] || {};
      matriz[am.classe][previsto] = (matriz[am.classe][previsto] || 0) + 1;
      classesVistas[am.classe] = true;
      if (previsto) classesVistas[previsto] = true;
    });

    /* Métricas por classe. Acurácia sozinha engana com classes
       desbalanceadas: um modelo que só responde "falha mecânica" pode
       acertar 40% e ser inútil. F1 por classe expõe isso. */
    var classes = Object.keys(classesVistas);
    var porClasse = classes.map(function (c) {
      var vp = (matriz[c] || {})[c] || 0;
      var fn = Object.keys(matriz[c] || {}).reduce(function (s, prev) {
        return s + (prev === c ? 0 : matriz[c][prev]);
      }, 0);
      var fp = classes.reduce(function (s, real) {
        if (real === c) return s;
        return s + ((matriz[real] || {})[c] || 0);
      }, 0);
      var precisao = (vp + fp) ? vp / (vp + fp) : null;
      var revocacao = (vp + fn) ? vp / (vp + fn) : null;
      var f1 = (precisao && revocacao) ? 2 * precisao * revocacao / (precisao + revocacao) : 0;
      return { classe: c, suporte: vp + fn, precisao: precisao, revocacao: revocacao, f1: f1 };
    }).sort(function (a, b) { return b.suporte - a.suporte; });

    var comSuporte = porClasse.filter(function (p) { return p.suporte > 0; });
    var f1Macro = comSuporte.length
      ? comSuporte.reduce(function (s, p) { return s + p.f1; }, 0) / comSuporte.length
      : 0;

    return {
      suficiente: true,
      modelo: modelo,
      atributos: atributos,
      totalAmostras: ordenadas.length,
      treino: treino.length,
      teste: teste.length,
      corteEm: teste.length ? teste[0].ts : null,
      periodoTreino: { inicio: treino[0].ts, fim: treino[treino.length - 1].ts },
      periodoTeste: teste.length ? { inicio: teste[0].ts, fim: teste[teste.length - 1].ts } : null,
      acuraciaTop1: teste.length ? acertosTop1 / teste.length : 0,
      acuraciaTop3: teste.length ? acertosTop3 / teste.length : 0,
      acuraciaBaseline: teste.length ? acertosBaseline / teste.length : 0,
      acuraciaBaselineTop3: teste.length ? acertosBaselineTop3 / teste.length : 0,
      f1Macro: f1Macro,
      matriz: matriz,
      classes: classes,
      porClasse: porClasse
    };
  }

  OEE.ACMP = {
    ATRIBUTOS_INICIO: ATRIBUTOS_INICIO,
    ATRIBUTOS_EM_CURSO: ATRIBUTOS_EM_CURSO,
    ROTULOS_ATRIBUTO: ROTULOS_ATRIBUTO,
    MINIMO_AMOSTRAS: MINIMO_AMOSTRAS,
    extrairAmostras: extrairAmostras,
    contextoAtual: contextoAtual,
    treinar: treinar,
    pontuar: pontuar,
    sugerir: sugerir,
    evidencias: evidencias,
    baseline: baseline,
    avaliar: avaliar,
    faixaHoraria: faixaHoraria,
    faixaDuracao: faixaDuracao
  };
})(window);

File: apps/api/src/oee/domain/acmp.py
"""Naive Bayes multinomial para ACMP — portado de legacy/acmp.js.

Contrato: sem vazamento pelo estado da máquina; duração só no modo em curso;
avaliação com split temporal.
"""

from __future__ import annotations

import math
from typing import Any

MINIMO_AMOSTRAS = 30
ALFA = 1.0
ATRIBUTOS_INICIO = ["maquina", "turno", "faixa_hora", "produto", "motivo_anterior"]
ATRIBUTOS_EM_CURSO = ATRIBUTOS_INICIO + ["faixa_duracao"]
ROTULOS_ATRIBUTO = {
    "maquina": "máquina",
    "turno": "turno",
    "faixa_hora": "faixa horária",
    "produto": "produto",
    "motivo_anterior": "parada anterior",
    "faixa_duracao": "duração",
}


def faixa_horaria(ts_ms: int) -> str:
    h = __import__("datetime").datetime.fromtimestamp(ts_ms / 1000).hour
    if h < 4:
        return "00h-04h"
    if h < 8:
        return "04h-08h"
    if h < 12:
        return "08h-12h"
    if h < 16:
        return "12h-16h"
    if h < 20:
        return "16h-20h"
    return "20h-24h"


def faixa_duracao(seg: int | None) -> str:
    if seg is None:
        return "desconhecida"
    if seg <= 60:
        return "até 1 min"
    if seg <= 300:
        return "até 5 min"
    if seg <= 900:
        return "até 15 min"
    if seg <= 3600:
        return "até 1 h"
    return "acima de 1 h"


def extrair_amostras(dataset: dict[str, Any], opcoes: dict[str, Any] | None = None) -> list[dict[str, Any]]:
    opcoes = opcoes or {}
    ordens = {o["id"]: o for o in dataset.get("ordens") or []}
    maquina_ids = opcoes.get("maquina_ids")
    paradas = [
        p
        for p in dataset.get("eventos_parada") or []
        if p.get("motivo_id")
        and (not maquina_ids or p.get("maquina_id") in maquina_ids)
        and (not opcoes.get("inicio") or p.get("inicio", 0) >= opcoes["inicio"])
        and (not opcoes.get("fim") or p.get("inicio", 0) <= opcoes["fim"])
    ]
    paradas.sort(key=lambda p: p.get("inicio") or 0)
    anterior_por_maquina: dict[str, str] = {}
    amostras = []
    for p in paradas:
        ordem = ordens.get(p.get("ordem_id"))
        amostras.append(
            {
                "id": p["id"],
                "ts": p.get("inicio"),
                "maquina_id": p.get("maquina_id"),
                "classe": p.get("motivo_id"),
                "atributos": {
                    "maquina": p.get("maquina_id"),
                    "turno": p.get("turno_id") or "sem-turno",
                    "faixa_hora": faixa_horaria(p.get("inicio") or 0),
                    "produto": ordem.get("produto_id") if ordem else "sem-ordem",
                    "motivo_anterior": anterior_por_maquina.get(p.get("maquina_id"), "nenhuma"),
                    "faixa_duracao": faixa_duracao(p.get("duracao_seg")),
                },
            }
        )
        anterior_por_maquina[p.get("maquina_id")] = p.get("motivo_id")
    return amostras


def contexto_atual(dataset: dict[str, Any], maquina: dict[str, Any], opcoes: dict[str, Any] | None = None) -> dict[str, Any]:
    opcoes = opcoes or {}
    agora = opcoes.get("agora") or int(__import__("time").time() * 1000)
    ordem = None
    if maquina.get("ordem_atual_id"):
        ordem = next((o for o in dataset.get("ordens") or [] if o["id"] == maquina["ordem_atual_id"]), None)
    anteriores = [
        p
        for p in dataset.get("eventos_parada") or []
        if p.get("maquina_id") == maquina["id"] and p.get("motivo_id") and (p.get("inicio") or 0) < agora
    ]
    anteriores.sort(key=lambda p: p.get("inicio") or 0)
    anterior = anteriores[-1] if anteriores else None
    from oee.domain.timeutil import turno_do_instante

    turno = turno_do_instante(dataset.get("turnos") or [], agora)
    return {
        "maquina": maquina["id"],
        "turno": turno.get("turno_id"),
        "faixa_hora": faixa_horaria(agora),
        "produto": ordem.get("produto_id") if ordem else "sem-ordem",
        "motivo_anterior": anterior.get("motivo_id") if anterior else "nenhuma",
        "faixa_duracao": faixa_duracao(opcoes.get("duracao_seg")),
    }


def treinar(amostras: list[dict[str, Any]], atributos: list[str] | None = None) -> dict[str, Any]:
    atributos = atributos or ATRIBUTOS_INICIO
    contagem_classe: dict[str, int] = {}
    contagem: dict[str, dict[str, dict[str, int]]] = {a: {} for a in atributos}
    valores: dict[str, dict[str, bool]] = {a: {} for a in atributos}
    total = 0
    for am in amostras:
        c = am["classe"]
        contagem_classe[c] = contagem_classe.get(c, 0) + 1
        total += 1
        for a in atributos:
            v = am["atributos"].get(a)
            contagem[a].setdefault(c, {})
            contagem[a][c][v] = contagem[a][c].get(v, 0) + 1
            valores[a][v] = True
    return {
        "tipo": "naive_bayes",
        "atributos": atributos,
        "classes": list(contagem_classe.keys()),
        "contagem_classe": contagem_classe,
        "contagem": contagem,
        "cardinalidade": {a: len(valores[a]) for a in atributos},
        "total": total,
    }


def pontuar(modelo: dict[str, Any], ctx: dict[str, Any]) -> list[dict[str, Any]]:
    if not modelo or not modelo.get("total"):
        return []
    pontos = []
    for c in modelo["classes"]:
        logp = math.log(modelo["contagem_classe"][c] / modelo["total"])
        for a in modelo["atributos"]:
            por_classe = modelo["contagem"][a].get(c) or {}
            n = por_classe.get(ctx.get(a), 0)
            total_classe = modelo["contagem_classe"][c]
            k = modelo["cardinalidade"].get(a) or 1
            logp += math.log((n + ALFA) / (total_classe + ALFA * k))
        pontos.append({"classe": c, "log": logp})
    maior = max(p["log"] for p in pontos)
    soma = 0.0
    for p in pontos:
        p["exp"] = math.exp(p["log"] - maior)
        soma += p["exp"]
    for p in pontos:
        p["prob"] = p["exp"] / soma if soma else 0.0
    return sorted(pontos, key=lambda p: p["prob"], reverse=True)


def evidencias(modelo: dict[str, Any], ctx: dict[str, Any], classe: str, limite: int = 2) -> list[dict[str, Any]]:
    lista = []
    for a in modelo["atributos"]:
        valor = ctx.get(a)
        dessa = (modelo["contagem"][a].get(classe) or {}).get(valor, 0)
        com_valor = sum((modelo["contagem"][a].get(c) or {}).get(valor, 0) for c in modelo["classes"])
        proporcao = (dessa / com_valor) if com_valor else 0.0
        rotulo = ROTULOS_ATRIBUTO.get(a, a)
        lista.append(
            {
                "atributo": a,
                "rotulo": rotulo,
                "valor": valor,
                "proporcao": proporcao,
                "suporte": com_valor,
                "ocorrencias": dessa,
                "frase": (
                    f"entre paradas com {rotulo} = {valor}, {dessa} de {com_valor} "
                    f"({proporcao:.0%}) foram deste motivo"
                    if com_valor
                    else f"{rotulo} = {valor}"
                ),
            }
        )
    lista = [e for e in lista if e["suporte"] >= 3 and e["proporcao"] > 0]
    lista.sort(key=lambda e: e["proporcao"], reverse=True)
    return lista[:limite]


def sugerir(modelo: dict[str, Any], ctx: dict[str, Any], quantidade: int = 3) -> list[dict[str, Any]]:
    if not modelo or modelo.get("total", 0) < MINIMO_AMOSTRAS:
        return []
    return [
        {
            "motivo_id": p["classe"],
            "probabilidade": p["prob"],
            "posicao": i + 1,
            "evidencias": evidencias(modelo, ctx, p["classe"], 2),
            "modelo": "naive_bayes",
        }
        for i, p in enumerate(pontuar(modelo, ctx)[:quantidade])
    ]


def baseline(amostras_treino: list[dict[str, Any]]) -> dict[str, Any]:
    por_maquina: dict[str, dict[str, int]] = {}
    geral: dict[str, int] = {}
    for a in amostras_treino:
        por_maquina.setdefault(a["maquina_id"], {})
        por_maquina[a["maquina_id"]][a["classe"]] = por_maquina[a["maquina_id"]].get(a["classe"], 0) + 1
        geral[a["classe"]] = geral.get(a["classe"], 0) + 1

    def ordenar(mapa: dict[str, int]) -> list[str]:
        return sorted(mapa.keys(), key=lambda k: mapa[k], reverse=True)

    geral_ord = ordenar(geral)
    return {
        "por_maquina": {m: ordenar(cnt)[0] if cnt else None for m, cnt in por_maquina.items()},
        "por_maquina_top3": {m: ordenar(cnt)[:3] for m, cnt in por_maquina.items()},
        "geral": geral_ord[0] if geral_ord else None,
        "geral_top3": geral_ord[:3],
    }


def avaliar(amostras: list[dict[str, Any]], opcoes: dict[str, Any] | None = None) -> dict[str, Any]:
    opcoes = opcoes or {}
    atributos = opcoes.get("atributos") or ATRIBUTOS_INICIO
    fracao = opcoes.get("fracao_treino", 0.75)
    if len(amostras) < MINIMO_AMOSTRAS * 2:
        return {"suficiente": False, "total_amostras": len(amostras), "minimo": MINIMO_AMOSTRAS * 2}
    ordenadas = sorted(amostras, key=lambda a: a["ts"] or 0)
    corte = int(len(ordenadas) * fracao)
    treino, teste = ordenadas[:corte], ordenadas[corte:]
    modelo = treinar(treino, atributos)
    ref = baseline(treino)
    acertos_top1 = acertos_top3 = acertos_base = acertos_base_top3 = 0
    matriz: dict[str, dict[str, int]] = {}
    classes_vistas: dict[str, bool] = {}
    for am in teste:
        ranking = pontuar(modelo, am["atributos"])
        previsto = ranking[0]["classe"] if ranking else None
        top3 = [p["classe"] for p in ranking[:3]]
        if previsto == am["classe"]:
            acertos_top1 += 1
        if am["classe"] in top3:
            acertos_top3 += 1
        if (ref["por_maquina"].get(am["maquina_id"]) or ref["geral"]) == am["classe"]:
            acertos_base += 1
        ref_top3 = ref["por_maquina_top3"].get(am["maquina_id"]) or ref["geral_top3"]
        if am["classe"] in ref_top3:
            acertos_base_top3 += 1
        matriz.setdefault(am["classe"], {})
        matriz[am["classe"]][previsto] = matriz[am["classe"]].get(previsto, 0) + 1
        classes_vistas[am["classe"]] = True
        if previsto:
            classes_vistas[previsto] = True
    classes = list(classes_vistas.keys())
    por_classe = []
    for c in classes:
        vp = (matriz.get(c) or {}).get(c, 0)
        fn = sum(v for k, v in (matriz.get(c) or {}).items() if k != c)
        fp = sum((matriz.get(real) or {}).get(c, 0) for real in classes if real != c)
        precisao = vp / (vp + fp) if (vp + fp) else None
        revocacao = vp / (vp + fn) if (vp + fn) else None
        f1 = (2 * precisao * revocacao / (precisao + revocacao)) if precisao and revocacao else 0.0
        por_classe.append({"classe": c, "suporte": vp + fn, "precisao": precisao, "revocacao": revocacao, "f1": f1})
    por_classe.sort(key=lambda p: p["suporte"], reverse=True)
    com = [p for p in por_classe if p["suporte"] > 0]
    f1_macro = sum(p["f1"] for p in com) / len(com) if com else 0.0
    n = len(teste) or 1
    return {
        "suficiente": True,
        "modelo": "naive_bayes",
        "atributos": atributos,
        "total_amostras": len(ordenadas),
        "treino": len(treino),
        "teste": len(teste),
        "acuracia_top1": acertos_top1 / n,
        "acuracia_top3": acertos_top3 / n,
        "acuracia_baseline": acertos_base / n,
        "acuracia_baseline_top3": acertos_base_top3 / n,
        "f1_macro": f1_macro,
        "matriz": matriz,
        "classes": classes,
        "por_classe": por_classe,
    }
